File: Edu-Kotlin/src/com/jetbrains/edu/kotlin/twitter/KtTwitterDialogPanel.java
package com.jetbrains.edu.kotlin.twitter;

import com.intellij.openapi.ui.VerticalFlowLayout;
import com.intellij.util.ui.UIUtil;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.twitter.TwitterUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.DocumentListener;
import java.awt.*;
import java.io.InputStream;
import java.net.URL;

@SuppressWarnings("WeakerAccess")
public class KtTwitterDialogPanel extends TwitterUtils.TwitterDialogPanel {
    private final JTextArea myTwitterTextField;
    private final JLabel myRemainSymbolsLabel;
    private URL myImageUrl;
    private String imageName = "";

    public KtTwitterDialogPanel(@NotNull Task solvedTask) {
        setLayout(new VerticalFlowLayout());
        myRemainSymbolsLabel = new JLabel();
        myTwitterTextField = new JTextArea();
        myTwitterTextField.setLineWrap(true);
        create(solvedTask);
    }

    public void create(@NotNull Task solvedTask) {
        add(new JLabel(UIUtil.toHtml("<b>Post your achievements to twitter!<b>\n")));
        myImageUrl = getMediaSourceForTask(solvedTask);
        addImageLabel();

        String messageForTask = getMessageForTask(solvedTask);
        myTwitterTextField.setText(messageForTask);
        add(myTwitterTextField);

        myRemainSymbolsLabel.setText(String.valueOf(140 - messageForTask.length()));
        JPanel jPanel = new JPanel(new BorderLayout());
        jPanel.add(myRemainSymbolsLabel, BorderLayout.EAST);
        add(jPanel);
    }

    private void addImageLabel() {
        if (myImageUrl != null) {
            Icon icon = new ImageIcon(myImageUrl);
            add(new JLabel(icon));
        }
    }

    @NotNull
    @Override
    public String getMessage() {
        return myTwitterTextField.getText();
    }

    private static String getMessageForTask(@NotNull final Task task) {
        int solvedTaskNumber = KtTwitterConfigurator.calculateTaskNumber(task);
        return "Hey, I just completed level " + solvedTaskNumber / 8 
                +" of Kotlin Koans. https://kotlinlang.org/docs/tutorials/koans.html #kotlinkoans";
    }

    @Nullable
    @Override
    public InputStream getMediaSource() {
        return getClass().getResourceAsStream(imageName);
    }

    @Nullable
    @Override
    public String getMediaExtension() {
        return "gif";
    }

    @Nullable
    private URL getMediaSourceForTask(@NotNull final Task task) {
        imageName = "/twitter/kotlin_koans/images/" + getImageName(task);
        return getClass().getResource(imageName);
    }
    
    private static String getImageName(@NotNull final Task task) {
        int solvedTaskNumber = KtTwitterConfigurator.calculateTaskNumber(task);
        int level = solvedTaskNumber / 8;
        return level + "level.gif";
    }


    @Override
    public void addTextFieldVerifier(@NotNull DocumentListener documentListener) {
        myTwitterTextField.getDocument().addDocumentListener(documentListener);
    }

    @NotNull
    @Override
    public JLabel getRemainSymbolsLabel() {
        return myRemainSymbolsLabel;
    }
}

File: Edu-Kotlin/src/com/jetbrains/edu/kotlin/twitter/KtTwitterDialogPanel.kt
package com.jetbrains.edu.kotlin.twitter

import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.ui.VerticalFlowLayout
import com.intellij.ui.components.JBScrollPane
import com.intellij.util.ui.JBUI
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.twitter.TwitterUtils.TwitterDialogPanel
import org.jetbrains.annotations.NonNls
import java.io.InputStream
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.JTextArea

class KtTwitterDialogPanel(private val solvedTask: Task) : TwitterDialogPanel(VerticalFlowLayout(0, 0)) {
  private val twitterTextField: JTextArea = JTextArea(4, 0)

  init {
    border = JBUI.Borders.empty()
    twitterTextField.lineWrap = true
    val messageForTask = getMessageForTask(solvedTask)
    twitterTextField.text = messageForTask
    // JTextArea doesn't support scrolling itself
    val scrollPane = JBScrollPane(twitterTextField)
    add(scrollPane)

    val imageUrl = javaClass.getResource(getImagePath(solvedTask))
    if (imageUrl != null) {
      val iconLabel = JLabel(ImageIcon(imageUrl))
      iconLabel.border = JBUI.Borders.empty(10, 0, 0, 0)
      add(iconLabel)
    }
  }

  override fun getMessage(): String = twitterTextField.text
  override fun getMediaSource(): InputStream? = javaClass.getResourceAsStream(getImagePath(solvedTask))
  override fun getMediaExtension(): String? = "gif"

  override fun doValidate(): ValidationInfo? {
    val extraCharacters = twitterTextField.text.length - 280
    return if (extraCharacters > 0) {
      ValidationInfo("Maximum length is 280 characters. Extra characters: ${extraCharacters}", twitterTextField)
    }
    else {
      super.doValidate()
    }
  }

  companion object {
    @NonNls
    private val COMPLETE_KOTLIN_KOANS_LEVEL = "Hey, I just completed level %d of Kotlin Koans. https://kotlinlang.org/docs/tutorials/koans.html #kotlinkoans"

    private fun getMessageForTask(task: Task): String {
      val solvedTaskNumber = KtTwitterConfigurator.calculateTaskNumber(task)
      return String.format(COMPLETE_KOTLIN_KOANS_LEVEL, solvedTaskNumber / 8)
    }

    private fun getImagePath(task: Task): String {
      val solvedTaskNumber = KtTwitterConfigurator.calculateTaskNumber(task)
      val level = solvedTaskNumber / 8
      return level.toString() + "level.gif"
    }
  }
}
